### ISSUE TITLE
Restrict pointer loads not affected by asm memory clobber

### ISSUE THREAD

**Issue by gooncreeper** [llvm/llvm-project#86734]
When a memory clobber is used in inline assembly, restrict pointers loads are incorrectly reordered after the inline assembly. [GCC inline assembly documentation](https://gcc.gnu.org/onlinedocs/gcc/extensions-to-the-c-language-family/how-to-use-inline-assembly-language-in-c-code.html#clobbers-and-scratch-registers) states

> Further, the compiler does not assume that any values read from memory before an asm remain unchanged after that asm

Therefore it is invalid to assume that values pointed by restrict pointers are unchanged if they are not passed as input operands.

Example C Code
```C
#include <stdint.h>

void ipc_copy(uint64_t p, uint64_t *restrict a, uint64_t *restrict b) {
    uint64_t s0 = a[0];
    uint64_t s1 = a[1];
    uint64_t s2 = a[2];
    uint64_t s3 = a[3];
    __asm__ __volatile__("mov %0, %%cr3" :: "r"(p) : "memory");
    b[0] = s0;
    b[1] = s1;
    b[2] = s2;
    b[3] = s3;
}
```
Clang
```ASM
ipc_copy:                               # @ipc_copy
        mov     cr3, rdi
        movups  xmm0, xmmword ptr [rsi]
        movups  xmmword ptr [rdx], xmm0
        movups  xmm0, xmmword ptr [rsi + 16]
        movups  xmmword ptr [rdx + 16], xmm0
        ret
```
Expected (Clang 14.0.0 / GCC)
```ASM
ipc_copy:                               # @ipc_copy
        movups  xmm0, xmmword ptr [rsi]
        movups  xmm1, xmmword ptr [rsi + 16]
        mov     cr3, rdi
        movups  xmmword ptr [rdx], xmm0
        movups  xmmword ptr [rdx + 16], xmm1
        ret
```

Related: https://github.com/llvm/llvm-project/issues/15867

**Comment by gooncreeper** [comment]
@Endilll I believe this should be labeled as a miscompilation.

**Comment by nikic** [comment]
I don't think this is a miscompilation. The asm block obviously cannot access any memory it does not have provenance to access (that would just completely break the entire memory model), so it cannot access a restrict pointer that was not passed to the asm (or escaped).

Similarly to what would happen with a local variable. A strict reading would imply that this optimization is invalid: https://c.godbolt.org/z/v7jr5rMs3 But of course it's intended to be valid. The wording in the GCC docs is just imprecise.
